mpqp/core/instruction/gates/controlled_gate.py

```diff
@@ -39,3 +41,118 @@ def __init__(
         """See parameter description."""
 
         Gate.__init__(self, targets, label)
+
+    def controlled_gate_to_matrix(self) -> Matrix:
+        """
+        Constructs the matrix representation of a controlled gate.
+
+        Returns:
+            The matrix representation of the controlled gate.
+        """
+        import numpy as np
+
+        if len(self.controls) != 1 or len(self.targets) != 1:
+            return self._multi_control_gate_to_matrix()
+        control, target = self.controls[0], self.targets[0]
+        min_ = min(control, target)
+        control -= min_
+        target -= min_
+
+        nb_qubit = abs(control - target) + 1
+        zero = np.array([[1, 0], [0, 0]])
+        one = np.array([[0, 0], [0, 1]])
+        I2 = np.eye(2)
+
+        control_matrix = zero if control == 0 else I2
+        target_matrix = (
+            one
+            if control == 0
+            else (self.non_controlled_gate.to_matrix() if target == 0 else I2)
+        )
+
+        for i in range(1, nb_qubit):
+            if i == control:
+                target_matrix = np.kron(target_matrix, one)
+            elif i == target:
+                target_matrix = np.kron(
+                    target_matrix, self.non_controlled_gate.to_matrix()
+                )
+            else:
+                target_matrix = np.kron(target_matrix, I2)
+
+        for i in range(1, nb_qubit):
+            if i == control:
+                control_matrix = np.kron(control_matrix, zero)
+            else:
+                control_matrix = np.kron(control_matrix, I2)
+
+        return control_matrix + target_matrix
+
+    def _multi_control_gate_to_matrix(self) -> Matrix:
```

Comment:
it sounds like the previous method is a special case of this one

Reply:
yes, when we have multiple control or target we use a different methode to compute. start from canonical matrix and do swap